### PR TITLE
fix(accounts): payment entry unallocated amount

### LINKED ISSUE #23936: Unallocated amount is incorrectly calculated in Payment Entry when paying an Invoice in Non Base currency
## Description of the issue
When a PI / SI is raised in a non Base currency (lets say USD) and we try to pay it through an account in the same none base currency (USD), the unallocated amount is wrongly calculated. Let me try to explain it in detail with an example. Lets say I have a supplier who deals in USD. The Supplier has raised me an invoice of 5000 USD and at the time book booking exchange rate w.r.t my currency was lets say 16. And at a later date I make a manual payment entry with an amount of 10000 USD ( paying from an account with currency USD ) and the exchange rate on that day was 15 which was entered in the PE for. Now unallocated amount in this scenario should be 5000 USD but the current implementation calculates it as 4333 USD. Now it calculates it incorrectly because it adds up all the allocated amount first in the base currency with exchange rate present in the PI and then calculates the unallocated amount in base currency and converts it back to USD with the exchange rate of PE but in reality, it shouldnt matter at all as the unallocated amount should just be 5000 USD as we are paying in USD to a PI raised in USD

## Context information (for bug reports)

**Output of `bench version`**
```
(paste here)
```

## Steps to reproduce the issue

1.
2.
3.

### Stacktrace / full error message

```
(paste here)
```

## Additional information

OS version / distribution, `ERPNext` install method, etc.

### LINKED ISSUE #29595: Unallocated amount is inaccurate on multi-currency Payment Entry
### Information about bug

The issue is that when you are processing a payment entry in a currency other than base, the unallocated_amount can end up being inaccurate, especially with higher exchange rates, e.g. 210 : 1.

In my example of the issue we are running a base currency of USD and we are receipting payment in ZWL. The customer has paid us ZWL$1,160,250.00 at an exchange rate of 210:1 (represented in ERPNext as 0.004761905) which evaluates to USD$5,525.00. This part is accurate, however because ERPNext does the allocations in base currency then converts back to the selected currency to give you your unallocated_amount, we get a value of ZWL$1,160,249.94. So we are losing ZWL$0.06 on the transaction but there is no USD difference because when converted it is less than USD$0.01. Thus there is a loss in value of the customer's money without the value going anywhere.

### Module

accounts

### Version

Frappe version - 13.19.0
ERPNext version - 13.19.0

### Installation method

manual install

### Relevant log output / Stack trace / Full Error Message.

_No response_

### PR DESCRIPTION
**Pull request by dj12djdjs**
Continuation of https://github.com/frappe/erpnext/pull/30504

closes: https://github.com/frappe/erpnext/issues/29595
closes: https://github.com/frappe/erpnext/issues/23936

** Allocating References
This shows the unallocated amount being changes as rows become allocated.

allocate references

** Write Offs
Using the "deductions" table, the unallocated_amount can be manipulated. It's primary use is for rounding, and writeoffs.

maniplute unallocated_amount

** Exchange gain/loss
Using the "losses table, the difference_amount can be manipulated. It's primary use if for declaring the difference between currencies. 
manipluate difference amount

### CODE REVIEW

File: erpnext/accounts/doctype/payment_entry/payment_entry.js [hunk 1]
```diff
@@ -871,16 +871,25 @@ frappe.ui.form.on('Payment Entry', {
...
 
 		if(frm.doc.party) {
-			if(frm.doc.payment_type == "Receive"
+			if(frm.doc.payment_type === "Receive"
 				&& frm.doc.base_total_allocated_amount < frm.doc.base_received_amount + total_deductions
 				&& frm.doc.total_allocated_amount < frm.doc.paid_amount + (total_deductions / frm.doc.source_exchange_rate)) {
-					unallocated_amount = (frm.doc.base_received_amount + total_deductions + frm.doc.base_total_taxes_and_charges
-						- frm.doc.base_total_allocated_amount) / frm.doc.source_exchange_rate;
-			} else if (frm.doc.payment_type == "Pay"
+					unallocated_amount = (
```
Comment (dj12djdjs) [inline erpnext/accounts/doctype/payment_entry/payment_entry.js:875]: Need to look into this again. Just found out Allocated credit notes is not working as expected

Comment (dj12djdjs) [inline erpnext/accounts/doctype/payment_entry/payment_entry.js:875]: Need to look into this. Allocating credit not working as expected

Comment (dj12djdjs) [inline erpnext/accounts/doctype/payment_entry/payment_entry.js:875]: Need to look into this. Allocating credit not working as expected

**Comment by dj12djdjs** [comment]
Disregard comment about allocation for credit note.

File: erpnext/accounts/doctype/payment_entry/payment_entry.js [hunk 2]
```diff
@@ -871,16 +871,25 @@ frappe.ui.form.on('Payment Entry', {
...
+						((frm.doc.base_received_amount + frm.doc.base_total_taxes_and_charges
+							 + (total_deductions > 0 ? total_deductions : 0))
+						/ frm.doc.source_exchange_rate)
+						- frm.doc.total_allocated_amount
+					);
+
+			} else if (frm.doc.payment_type === "Pay"
 				&& frm.doc.base_total_allocated_amount < frm.doc.base_paid_amount - total_deductions
```
Comment (dj12djdjs) [inline erpnext/accounts/doctype/payment_entry/payment_entry.js]: This is still some instance where this condition results in unallocated_amount = 0 even though there is still a balance remaining to be allocated.

**Comment by dj12djdjs** [comment]
I'm starting to think the only way to solve this is to separate deductions/writeoffs and exchange rate deductions into separate  child tables.

Having a condition that forces unallocated_amount to be zero is not working. It should be produced by addition and subtraction.

The only way I can think to solve this is to subtract exchange deductions from total_deductions. Then calculate the unallocated_amount

**Comment by dj12djdjs** [comment]
Gonna post some screen recordings in a few hours.

**Comment by dj12djdjs** [comment]
I don't know why the linter keeps failing. I've run `pre-commit run --all-files` multiple times and is passing on my machine.

**Comment by dj12djdjs** [comment]
@deepeshgarg007 This is working. Can you please review.

**Comment by dj12djdjs** [comment]
@deepeshgarg007 Just checking up, did you have a chance to review this ?

**Comment by deepeshgarg007** [comment]
@dj12djdjs I am reviewing this and will also take a second opinion internally if we need a separate table for losses

**Comment by deepeshgarg007** [comment]
@dj12djdjs I am reviewing this and will also take a second opinion internally if we need a separate table for losses

**Comment by dj12djdjs** [comment]
> @dj12djdjs I am reviewing this and will also take a second opinion internally if we need a separate table for losses

If you guys can find a way to make it work in 1 table that would be more convenient.
 
I don't have much experience in accounting. This is the only way I could make it work.

**Comment by dj12djdjs** [comment]
@deepeshgarg007 any update from the team on this matter?

**Comment by dj12djdjs** [comment]
@deepeshgarg007 Is there a time frame for this?

**Comment by OneiricIT** [comment]
@dj12djdjs does this have any relation to being able to pull credit note amounts sitting on account as payment method for other invoices?
Reference:
https://github.com/frappe/erpnext/pull/28775

**Comment by dj12djdjs** [comment]
> @dj12djdjs does this have any relation to being able to pull credit note amounts sitting on account as payment method for other invoices? Reference: #28775

@OneiricIT no, this PR is just about unallocated calculation method

**Comment by deepeshgarg007** [comment]
Closing this since a new [PR](https://github.com/frappe/erpnext/pull/31272) is raised for this.

**Pull request closed without merging**